### PR TITLE
#271: Add support for basic entities

### PR DESCRIPTION
**Pull request by rzo1**
# What does this PR do?

- Adds support for basic entities (hashtags, cashtags, mentions, urls) in V1 and V2 tweet objects.

### CODE REVIEW

File: src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV2.java [hunk 1]
```diff
@@ -221,7 +239,8 @@ public TweetType getTweetType() {
     private String                   inReplyToUserId;
     @JsonProperty("referenced_tweets")
     private List<ReferencedTweetDTO> referencedTweets;
-    private JsonNode                 entities;
+    @JsonProperty("entities")
```
Comment (redouane59) [inline src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV2.java:242]: this line is useless no ?

**Review by redouane59 — COMMENTED** [review]
Awesome job, just let you one little comment :)

File: src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV2.java (hunk 1 above)
Comment (rzo1) [reply src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV2.java:242]: As long as we do not rename the class attribute `entities`, the annotation can be removed :)

File: src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java [hunk 2]
```diff
@@ -100,6 +109,11 @@ public String getSource() {
     return null;
   }
 
+  @Override
+  public Entities getEntities() {
```
Comment (redouane59) [inline src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java:113]: Just one doubt, can't it be managed natively thanks to Lombok thanks to `@Getter` annotation ?

File: src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java [hunk 3]
```diff
@@ -112,4 +126,132 @@ public String getAuthorId() {
...
+    private List<UserMentionEntityV1> userMentions;
+    @JsonProperty("symbols")
+    private List<SymbolEntityV1> symbols;
+    private JsonNode media; //TODO We should implement media objects accordingly
+  }
+
+  @Getter
+  @Setter
+  public static class BaseEntityV1 implements BaseEntity {
+    @JsonProperty("indices")
```
Comment (redouane59) [inline src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java:148]: also useless :)

**Comment by rzo1** [comment]
Sorry - had to push an additional fix.

Looks like https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/tweet is wrong as it states `tag` for `mentions` instead of `username` but returns `username` in fact...

File: src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java [hunk 4]
```diff
@@ -112,4 +126,132 @@ public String getAuthorId() {
...
+    @Override
+    public int getEnd() {
+      return indices[1];
+    }
+  }
+
+  @Getter
+  @Setter
+  public static class TextBaseEntityV1 extends BaseEntityV1 implements TextBaseEntity {
+    @JsonProperty("text")
```
Comment (redouane59) [inline src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java:165]: also useless

File: src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java (hunk 2 above)
Comment (rzo1) [reply src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java:113]: Yup.

File: src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java [hunk 5]
```diff
@@ -112,4 +126,132 @@ public String getAuthorId() {
...
+    @Override
+    public String getText() {
+      return text;
+    }
+  }
+
+  @Getter
+  @Setter
+  public static class UrlEntityV1 extends BaseEntityV1 implements UrlEntity {
+    @JsonProperty("url")
```
Comment (redouane59) [inline src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java:177]: useless too

File: src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java [hunk 6]
```diff
@@ -112,4 +126,132 @@ public String getAuthorId() {
...
+    private String url;
+    @JsonProperty("display_url")
+    private String displayUrl;
+    @JsonProperty("expanded_url")
+    private String expandedUrl;
+    @JsonProperty("unwound")
```
Comment (redouane59) [inline src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java:183]: same

File: src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java [hunk 7]
```diff
@@ -112,4 +126,132 @@ public String getAuthorId() {
...
+      if(unwound != null) {
+        return unwound.getUrl();
+      }
+      return null;
+    }
+  }
+
+  @Getter
+  @Setter
+  public static class UnwoundUrlEntity {
```
Comment (redouane59) [inline src/main/java/io/github/redouane59/twitter/dto/tweet/TweetV1.java:221]: same on this class

**Comment by redouane59** [comment]
Finally they are a lot of not needed Jackson annotation, if you can just remove them and I'll merge the PR right after :)
Thanks !

**Comment by rzo1** [comment]
Shall I rebase the commit or squash it before?

**Comment by redouane59** [comment]
> Shall I rebase the commits or squash it before?

don't worry it's OK like that. I'll rebase & merge it after the build success :)

Good job !

**Pull request merged**